### ISSUE TITLE
How we can run Llama-2 in a low spec GPU? 6GB VRAM

### ISSUE THREAD

**Issue by fabiomb** [facebookresearch/llama#381]
As many of us I don´t have a huge CPU available but I do have enogh RAM, even with it´s limitations, it´s even possible to run Llama on a small GPU? RTX 3060 with 6GB VRAM here.

Of course i got the usual error:

`File "/usr/local/lib/python3.10/dist-packages/fairscale/nn/model_parallel/layers.py", line 262, in __init__
    self.weight = Parameter(torch.Tensor(self.output_size_per_partition, self.in_features))
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 86.00 MiB (GPU 0; 6.00 GiB total capacity; 5.28 GiB already allocated; 0 bytes free; 5.29 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 0 (pid: 9010) of binary: /usr/bin/python3`

and i know is just the first day until we can get some documentation for this kind of situation, but probably someone did the job with Llama-1 and is not as hard as just parameters (I Hope)

I only want to run the example text completion

torchrun --nproc_per_node 1 example_text_completion.py \
    --ckpt_dir llama-2-7b/ \
    --tokenizer_path tokenizer.model \
    --max_seq_len 128 --max_batch_size 4

Can i use the VRAM and RAM at the same time?

**Comment by J50** [comment]
might be able to try bitesandbytes:
https://colab.research.google.com/drive/1ge2F1QSK8Q7h0hn3YKuBCOAS0bK8E0wf?usp=sharing#scrollTo=582BktNnz76l

**Comment by tomorrmato** [comment]
Have you tried deepspeed bit 4 inference yet?

**Comment by jonmach** [comment]
I have the same problem.

I'm using a 1080Ti 11GB card, and it won't even run the samples outlined.  We don't all have 32GB cards lying around.

Is there any way to configure this to use swap. It will be slow, but it will come back eventually.

**Comment by fabiomb** [comment]
no, still very complicated to change to 4 bit without proper tutorials and guidelines, not an expert here

as far as i undertand if we convert from 8 bit to 4 bit it´s possible to use it, i'm right?

**Comment by jonmach** [comment]
I've now taken a different approach and instead of using the Llama-2 sample code, I switched to the llama.ccp codeset, and converted the base models directly into 4-bits.  There is a convert.py and quantise executables to convert this to ggml format. This worked flawlessly.

**Comment by Revanthmk** [comment]
> I've now taken a different approach and instead of using the Llama-2 sample code, I switched to the llama.ccp codeset (https://github.com/ggerganov/llama.cpp), and converted the base models directly into 4-bits. There is a convert.py and quantise executables (within the repository) to convert this to ggml format. This worked flawlessly.

Is there any way to use the 8bit or 4bit version directly from the main repo?

**Comment by jonmach** [comment]
You can download many ggml models from huggingface:

https://huggingface.co/models?search=ggml

**Comment by Revanthmk** [comment]
> You can download many ggml models from huggingface:
> 
> https://huggingface.co/models?search=ggml

So can I download the model from hugging face and use this repo to use that model for inference?

**Comment by jonmach** [comment]
I can't comment on whether it would work perfectly.  I'm just saying that there are a number of models in hugging face claiming to be ggml versions of the original llama weights.  My personal preference is to build them myself using the llama.cpp code (convert.py and quantize).  I'm then responsible for the results and makes my personal debugging and episodes of confusion much clearer.

It takes minutes to convert them

**Comment by Revanthmk** [comment]
> I can't comment on whether it would work perfectly. I'm just saying that there are a number of models in hugging face claiming to be ggml versions of the original llama weights. My personal preference is to build them myself using the llama.cpp code (convert.py and quantize). I'm then responsible for the results and makes my personal debugging and episodes of confusion much clearer.
> 
> It takes minutes to convert them

You can convert it using llama.cpp code? will check it out thanks.
